T-twr plus open edition support

nmuntyanov opened the pull request:
Hello.

THIS PR IS NOT PRODUCTION READY


I've found your project and interested in a lot.
I have T-TWR open edition VHF version 

So I took this https://github.com/OpenRTX/sa8x8-fw firmware, and wrote an abstraction layer for SA868 module
this library (https://github.com/OpenRTX/OpenRTX/blob/master/platform/drivers/baseband/AT1846S_SA8x8.cpp) was taken as a reference

I've tested APRS tracker, voice tx/rx

So if you interested in this, we can try to extend it and make it fully workable

lyusupov commented:
@nmuntyanov you may consider to contribute your efforts _(if this PR merge will fail)_ into an alternative ESP32-S3 APRS project: https://github.com/Hamlabs/ArcticTracker-ESP32

**ArcticTracker-ESP32** is a mature project (4 years old). It uses a hardware which is quite similar to the T-TWR Plus.
The  **ArcticTracker-ESP32** author is friendly for porting the project on the T-TWR Plus as well:
1) https://github.com/Hamlabs/ArcticTracker-ESP32/issues/3
2) https://github.com/Hamlabs/ArcticTracker-ESP32?tab=readme-ov-file#future-work

<img width="607" alt="image" src="https://github.com/nakhonthai/ESP32APRS_T-TWR/assets/5849637/7a8171c2-7a1f-4697-a88b-cecd867756cc">

lyusupov commented:
@nmuntyanov 
use of exceptions (try...catch...)  is a bad practice for Arduino/PlatformIO compliant libraries/sketches.
https://forum.arduino.cc/t/try-catch/180063

Would you mind to alter your code with no use of exceptions ?

nmuntyanov commented on platformio.ini:
```diff
@@ -35,5 +35,12 @@ build_flags =
+[env:sa868_vhf_open_edition]
+build_flags= ${env.build_flags} -DSA868_OPEN_EDITION
+
+[env:sa868_vhf_open_edition_debug]
```
@lyusupov maybe its better to do something like this?
separate build for each module?

nmuntyanov commented:
@lyusupov 
> @nmuntyanov use of exceptions (try...catch...) is a bad practice for Arduino/PlatformIO compliant libraries/sketches. https://forum.arduino.cc/t/try-catch/180063
> 
> Would you mind to alter your code with no use of exceptions ?

try-catch blocks was removed

nmuntyanov commented on src/sa868_open_edition.cpp:
```diff
@@ -0,0 +1,390 @@
+    String result;
+    sprintf(str, "AT+POKE=%d,%d\r\n", reg, value);
+    if (!SA868_WaitResponse(SerialRF, str, &result)) {
+        ESP_LOGD("SA8x8", "Error: reg: %02X <- val: %02X", reg, value);
```
```suggestion
        ESP_LOGE("SA8x8", "Error: reg: %02X <- val: %02X", reg, value);
```

lyusupov commented:
@nmuntyanov 

> try-catch blocks was removed

LGTM, thank you!

> separate build for each module?

IMHO, to split one firmware binary on two, _(radio specific)_ ones - it is a good start !

However, if you will eventually decide to support both NiceRF and OpenEdition drivers **within a single ESP32-S3 firmware binary** by having 2 instances of 'class SA868' and differentiate  them by executing Version()  call in runtime during the SA868 module probing sequence  - that would be perfect !

lyusupov commented:
@nmuntyanov 
I've pulled current 'master' of  the ESP32APRS_T-TWR repo and rebult -

```
======================================== [SUCCESS] Took 87.03 seconds ========================================

Environment                   Status    Duration
----------------------------  --------  ------------
sa868_vhf_open_edition        SUCCESS   00:02:49.842
sa868_vhf_open_edition_debug  SUCCESS   00:01:27.028
========================================= 2 succeeded in 00:04:16.870 =========================================

```

_My question is_: no PIO **build for SA868 with NiceRF firmware** anymore ?

lyusupov commented:
@nmuntyanov 
I've made an attempt to use the firmware with Open Edition module.

![](https://raw.githubusercontent.com/lyusupov/SoftRF/36e07ddbd14d4c18e61b59d254f3a4bc6fc21378/documents/images/ham-27.jpg)

<br>

### Issue No.1  
TRACKER function is enabled by default. It is worth to turn it back off.

<img width="659" alt="image" src="https://github.com/nakhonthai/ESP32APRS_T-TWR/assets/5849637/449922aa-41b2-48bc-bd8f-292d9ab803b5">

<br>
<br>

### Issue No.2
Missing radio module info and unable to store settings.

<img width="697" alt="image" src="https://github.com/nakhonthai/ESP32APRS_T-TWR/assets/5849637/c03af113-0d97-446d-bfa4-526d005e4a21">

nmuntyanov commented:
@lyusupov 
it was draft pr, not production ready, I have added a notice in description.
I have just implemented driver for openEdition
no backward compatibility yet.

nakhonthai commented:
Thank @nmuntyanov for code sa8x8.
I'm developing and testing.
![462599](https://github.com/nakhonthai/ESP32APRS_T-TWR/assets/16043758/7ef9d39f-16ca-41da-bad4-b0c0129e6df1)
![messageImage_1706761421682](https://github.com/nakhonthai/ESP32APRS_T-TWR/assets/16043758/f8e7dd12-b9c9-49d2-867b-0e56c170fef0)
![messageImage_1706761362047](https://github.com/nakhonthai/ESP32APRS_T-TWR/assets/16043758/0de1dee6-1d3d-48d7-9e0a-f9eaf5171592)
